### PR TITLE
icub3: add complete joint test

### PR DESCRIPTION
**Pull request by Nicogene**
It fixes https://github.com/robotology/icub-models-generator/issues/179.

It also fixes partially https://github.com/robotology/icub-models-generator/issues/181

cc @GiulioRomualdi

### CODE REVIEW

File: tests/icub-model-test.cpp [hunk 1]
```diff
@@ -328,15 +406,44 @@ bool checkFTSensorsAreCorrectlyOriented(iDynTree::KinDynComputations & comp)
...
-    ok = ok && checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_arm_ft_sensor");
-    ok = ok && checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "l_leg_ft_sensor");
-    ok = ok && checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_leg_ft_sensor");
-    ok = ok && (checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "l_foot_ft_sensor") || isPlusModel);
-    ok = ok && (checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_foot_ft_sensor") || isPlusModel);
+    ok = ok & checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_arm_ft_sensor");
+    ok = ok & checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "l_leg_ft_sensor");
+    ok = ok & checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_leg_ft_sensor");
+    ok = ok & (checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "l_foot_ft_sensor") || isPlusModel);
+    ok = ok & (checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_foot_ft_sensor") || isPlusModel);
```
Comment (traversaro) [inline tests/icub-model-test.cpp:413]: Not a problem or blocking PR, but please avoid unrelated changes in PRs, especially changes like this that change the use of the logical operator `&&` to the bitwise operator `&` , that in this case should work, but in some cases could lead to unexpected results (see Point 3 of https://www.geeksforgeeks.org/bitwise-operators-in-c-cpp/).

Comment (Nicogene) [reply tests/icub-model-test.cpp:413]: I put `&` instead of `&&` in order to check all the fts instead of stopping at the first failure, I can change it  keeping `&&` and inverting the operands

File: tests/icub-model-test.cpp [hunk 2]
```diff
@@ -259,6 +259,84 @@ bool checkAxisDirections(iDynTree::KinDynComputations & comp)
...
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(0.267012,-0.960459,0.0788901));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(-0.0713662,0.0619303,0.995526));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(0.961047,0.271447,0.0520081));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(-0.267012,0.960459,-0.0788901));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(0.961047,-0.271447,0.0520081));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(0.0713662,0.0619303,-0.995526));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(-0.267012,-0.960459,-0.0788901));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(0.0713662,0.0619303,-0.995526));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(-0.961047,0.271447,-0.0520081));
+    expectedDirectionInRootLink.push_back(iDynTree::Direction(0.267012,0.960459,0.0788901));
```
Comment (traversaro) [inline tests/icub-model-test.cpp:316]: Sorry for being picky, but the code as written in this form would be quite complicate to maintain, as it is difficult to understand which axis correspond to which joint. 

Could it make sense to write it as: 
~~~
    axisNames.push_back("torso_roll");
    expectedDirectionInRootLink.push_back(iDynTree::Direction(-1,0,0));
    axisNames.push_back("l_hip_pitch");
    expectedDirectionInRootLink.push_back(iDynTree::Direction(0,-1,0));
    axisNames.push_back("r_hip_pitch");
    expectedDirectionInRootLink.push_back(iDynTree::Direction(0,-1,0));
~~~

And so on so forth?

File: tests/icub-model-test.cpp [hunk 3]
```diff
@@ -328,15 +406,44 @@ bool checkFTSensorsAreCorrectlyOriented(iDynTree::KinDynComputations & comp)
...
+
+    iDynTree::Rotation rootLink_R_sensorFrameExpectedFoot =
+        iDynTree::Rotation(-0.5, 0.866025, 0,
+                           -0.866025, -0.5, 0,
+                            0, 0, 1);
+
+    bool ok = checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameLeftArmExpected, "l_arm_ft_sensor");
+    ok = ok & checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_arm_ft_sensor");
+    ok = ok & checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "l_leg_ft_sensor");
+    ok = ok & checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_leg_ft_sensor");
```
Comment (traversaro) [inline tests/icub-model-test.cpp:439]: This is extremely strange. I would expect the FT frames of the FT in the legs to have the Z direction corresponding to the root link, and then having just an additional rotation in the Z direction due to how the sensor is mounted, as it happens in the feet. As we did not validated the FT frames in the model, this may suggest that the frame orientation is not correct, so for the time being I would avoid putting them in the tests if we do not know if this frames are correct.

**Review by traversaro — CHANGES_REQUESTED** [review]
See inline comments.

File: tests/icub-model-test.cpp (hunk 1 above)
Comment (traversaro) [reply tests/icub-model-test.cpp:413]: > I put `&` instead of `&&` in order to check all the fts instead of stopping at the first failure, I can change it keeping `&&` and inverting the operands

Ahhhh, I am afraid this is not clear at all, and is kind of abusing that `&` is. I suggest to do something like:
~~~
    sensCheck = checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "r_arm_ft_sensor");
    ok = ok && sensCheck;
    sensCheck = checkFTSensorIsCorrectlyOriented(comp, rootLink_R_sensorFrameExpected, "l_leg_ft_sensor");
    ok = ok && sensCheck;
~~~

Note that if we want this kind of functionalities (i.e. the difference between `REQUIRE` and `CHECK` in catch, see https://github.com/catchorg/Catch2/blob/devel/docs/assertions.md#natural-expressions) probably it could make sense to use for this a  proper testing framework (the aforementioned catch2, for example).

File: tests/icub-model-test.cpp (hunk 2 above)
Comment (Nicogene) [reply tests/icub-model-test.cpp:332]: Now should be clear, it would be better to creat a struct with name and expected rotation but it may be refactored in the future

File: tests/icub-model-test.cpp (hunk 3 above)
Comment (Nicogene) [reply tests/icub-model-test.cpp]: I agree, moreover now the fts in the legs are disabled, then I remove it.

**Review by traversaro — APPROVED** [review]
Perfect, thanks!

**Comment by traversaro** [comment]
Feel free to merge in the form you prefer (squash/no squash).
